### ISSUE TITLE
ML Projects

### ISSUE THREAD

**Issue by susantkumarsahoo** [susantkumarsahoo/Car-Price-Prediction#21]
# US - VISA-PREDICTION

```python

# ==============================
# 1. Import Libraries
# ==============================
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import warnings
from datetime import date
import pickle

warnings.filterwarnings("ignore")
pd.set_option("display.max_columns", None)


# ==============================
# 2. Load Dataset
# ==============================
df = pd.read_csv("Visadataset.csv")
print("Dataset Shape:", df.shape)


# ==============================
# 3. Data Cleaning
# ==============================
# Missing values check
features_with_na = [col for col in df.columns if df[col].isnull().sum() > 0]
for feature in features_with_na:
    print(f"{feature} : {np.round(df[feature].isnull().mean()*100, 5)}% missing values")

# Drop unwanted columns
df.drop("case_id", axis=1, inplace=True)

# Calculate company age
current_year = date.today().year
df["company_age"] = current_year - df["yr_of_estab"]
df.drop("yr_of_estab", axis=1, inplace=True)


# ==============================
# 4. Feature Engineering
# ==============================
# Numerical & categorical features
num_features = [col for col in df.columns if df[col].dtype != "O"]
cat_features = [col for col in df.columns if df[col].dtype == "O"]

print("Numerical Features:", len(num_features))
print("Categorical Features:", len(cat_features))

# Split features & target
X = df.drop("case_status", axis=1)
y = np.where(df["case_status"] == "Denied", 1, 0)


# ==============================
# 5. Preprocessing Pipeline
# ==============================
from sklearn.preprocessing import OneHotEncoder, StandardScaler, OrdinalEncoder, PowerTransformer
from sklearn.compose import ColumnTransformer
from sklearn.pipeline import Pipeline

# Define transformations
or_columns = ["has_job_experience", "requires_job_training", "full_time_position", "education_of_employee"]
oh_columns = ["continent", "unit_of_wage", "region_of_employment"]
transform_columns = ["no_of_employees", "company_age"]

numeric_transformer = StandardScaler()
oh_transformer = OneHotEncoder()
ordinal_encoder = OrdinalEncoder()
transform_pipe = Pipeline(steps=[("transformer", PowerTransformer(method="yeo-johnson"))])

preprocessor = ColumnTransformer([
    ("OneHotEncoder", oh_transformer, oh_columns),
    ("OrdinalEncoder", ordinal_encoder, or_columns),
    ("Transformer", transform_pipe, transform_columns),
    ("StandardScaler", numeric_transformer, num_features)
])

# Fit-transform preprocessing
X = preprocessor.fit_transform(X)

# Save preprocessing pipeline
with open("preprocessor.pkl", "wb") as file:
    pickle.dump(preprocessor, file)


# ==============================
# 6. Handle Imbalance
# ==============================
from imblearn.combine import SMOTEENN
X_res, y_res = SMOTEENN(random_state=42, sampling_strategy="minority").fit_resample(X, y)


# ==============================
# 7. Train-Test Split
# ==============================
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X_res, y_res, test_size=0.2, random_state=42)


# ==============================
# 8. Model Training (Best Model Selection)
# ==============================
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import accuracy_score, classification_report

# Load tuned params (example assumes KNN is best)
best_knn_params = {"algorithm": "auto", "weights": "uniform", "n_neighbors": 5}
best_model = KNeighborsClassifier(**best_knn_params)

# Train model
best_model.fit(X_train, y_train)

# Evaluate
y_pred = best_model.predict(X_test)
print("FINAL MODEL: KNN")
print("Accuracy Score:", accuracy_score(y_test, y_pred))
print(classification_report(y_test, y_pred))

# Save trained model
with open("final_knn_model.pkl", "wb") as file:
    pickle.dump(best_model, file)


# ==============================
# 9. User Prediction (Form Input)
# ==============================
# Load saved model & preprocessor
with open("final_knn_model.pkl", "rb") as file:
    model = pickle.load(file)

with open("preprocessor.pkl", "rb") as file:
    preprocessor = pickle.load(file)

# Example User Input (from form)
user_data = {
    "continent": ["Asia"],
    "education_of_employee": ["Master's"],
    "has_job_experience": ["Y"],
    "requires_job_training": ["N"],
    "no_of_employees": [20000],
    "region_of_employment": ["West"],
    "prevailing_wage": [5500],
    "unit_of_wage": ["Hour"],
    "full_time_position": ["Y"],
    "company_age": [50]
}

# Convert to DataFrame
user_df = pd.DataFrame(user_data)

# Apply preprocessing
X_processed = preprocessor.transform(user_df)

# Predict
prediction = model.predict(X_processed)

# Output
print("Visa Status:", "Denied ❌" if prediction[0] == 1 else "Approved ✅")
